theano/gof/tests/test_opt.py

```diff
@@ -317,6 +317,18 @@ def test_straightforward(self):
         MergeOptimizer().optimize(g)
         assert str(g) == "[Op1(*1 -> Op2(x, y), *1, Op2(x, z))]"
 
+    def test_nondeterministic(self):
+        x, y, z = inputs()
+        op2.nondeterministic = True
```

Comment:
What is the goal of that?
Currently, Theano op are all pseudo-deterministic for a fixed input.
Can you describe more your use case?
For random number generator, we put the seed as an input of the node to keep that pseudo-deterministic behavior.

Maybe there is a an existing mechanism that would allow that.

Reply:
As I mentioned in the [initial comment](https://github.com/Theano/Theano/pull/6691#issue-253766114), `MergeOptimizer` will merge nodes for `Op`s with the same input, which causes problems for `Op`s that are fundamentally non-deterministic (e.g. by erroneously merging/removing them).  Also, I couldn't find a better means of addressing this at the `Op` or `FunctionGraph` level.

Otherwise, working around this design limitation by pushing the expression/enforcement of non-determinism to the input level only relocates the problem (and/or attempts to make it user-level).  That's all perfectly fine if Theano is expressly designed to have a deterministic `Op` restriction, but, when a direct solution for (at least some) `Op`-level non-determinism could be as simple as this PR, I think it's worth reconsidering.

Furthermore, the problem appears for `RandomFunction` when one uses a shared RNG state:
```python
import numpy as np
import theano
import theano.tensor as tt
from theano.printing import debugprint as tt_dprint
from theano.gof.fg import FunctionGraph
from theano.gof.opt import MergeOptimizer

rand_state = theano.shared(np.random.RandomState())
rvop = lambda x, y: tt.raw_random.normal(rand_state, [1], x, y)[1]

x, y, z = tt.vectors('xyz')
e = tt.add(rvop(x, y), rvop(x, y), rvop(x, z))
g = FunctionGraph([rand_state, x, y, z], [e])
g_opt = g.clone()

MergeOptimizer().optimize(g_opt)
tt_dprint([g, g_opt])
```
```
Elemwise{add,no_inplace} [id A] ''   9
 |RandomFunction{normal}.1 [id B] ''   8
 | |<RandomStateType> [id C]
 | |Elemwise{Cast{int64}} [id D] ''   7
 | | |MakeVector{dtype='int8'} [id E] ''   6
 | |   |TensorConstant{1} [id F]
 | |x [id G]
 | |y [id H]
 |RandomFunction{normal}.1 [id I] ''   5
 | |<RandomStateType> [id C]
 | |Elemwise{Cast{int64}} [id J] ''   4
 | | |MakeVector{dtype='int8'} [id K] ''   3
 | |   |TensorConstant{1} [id F]
 | |x [id G]
 | |y [id H]
 |RandomFunction{normal}.1 [id L] ''   2
   |<RandomStateType> [id C]
   |Elemwise{Cast{int64}} [id M] ''   1
   | |MakeVector{dtype='int8'} [id N] ''   0
   |   |TensorConstant{1} [id F]
   |x [id G]
   |z [id O]
Elemwise{add,no_inplace} [id P] ''   4
 |RandomFunction{normal}.1 [id Q] ''   3
 | |<RandomStateType> [id R]
 | |Elemwise{Cast{int64}} [id S] ''   1
 | | |MakeVector{dtype='int8'} [id T] ''   0
 | |   |TensorConstant{1} [id U]
 | |x [id V]
 | |y [id W]
 |RandomFunction{normal}.1 [id Q] ''   3
 |RandomFunction{normal}.1 [id X] ''   2
   |<RandomStateType> [id R]
   |Elemwise{Cast{int64}} [id S] ''   1
   |x [id V]
   |z [id Y]
```

Now, I'm guessing that the intended use is something like the following (because it doesn't erroneously merge/remove nodes):
```python
rv1 = rvop(x, y, rand_state)
rv2 = rvop(x, y, rv1[0])
rv3 = rvop(x, z, rv2[0])
e = tt.add(rv1[1], rv2[1], rv3[1])
```
Unfortunately, whether that is the intended use or not, we've already demonstrated how the design problem persists even at the input/user level&mdash;e.g. by introducing errors through otherwise acceptable use of the interface and/or sacrificing the basic interface for something much more cumbersome.  While this specific example could potentially be fixed by special considerations in the relevant `Op`s, `MergeFeature`, etc., or with additional documentation, both are clearly undesirable kludges that leave the underlying problems intact.

Reply:
That merge is not erroneous or incorrect, though. It will give the same numeric answer as the un-optimized version.
The NumPy code equivalent to "rvop" would be something like:
```
def rvop(x, y):
    r_state = copy.copy(rand_state)
    sample = r_state.normal(x, y)
    return sample
```

You can see the random state as a monad, that needs to be passed as input and returned as output with a different value.
In that sense, an op with the same inputs (including the state) will always return the same output, so they can be merged.
This is important because for some gradient computations, for instance, we may need the same sequence of random numbers to be generated more than once.

If you are proposing that, in some circumstances, we may want to replace one random state with another as the input of a sampling node, or merge sampling nodes that have the same inputs except from the random state, there is certainly a use for such an "unsafe" optimization. For instance, dead code elimination could be allowed to remove sampling nodes that only update the state if the sample is never used.
However I do not believe that the default, regular way of sampling numbers should become non-deterministic, when currently it is not.

Reply:
> That merge is not erroneous or incorrect, though. It will give the same numeric answer as the un-optimized version.

That's only true if the underlying RNG state happens to get copied at some point at or before the implementation/`perform` level, but it is dangerously incorrect at the graph level.  If Theano is designed to operate on that assumption, then it's setting an unnecessary and error-prone requirement/dependency between the graph and compute-levels&mdash;or its graphs are providing an extremely misleading representation of the relevant operations.  

> You can see the random state as a monad, that needs to be passed as input and returned as output with a different value.
> In that sense, an op with the same inputs (including the state) will always return the same output, so they can be merged.
> This is important because for some gradient computations, for instance, we may need the same sequence of random numbers to be generated more than once.

I understand what `RandomStreams` is trying to accomplish/model and the desire/need for side-effect-free operations, determinism, etc., but that's not clearly&mdash;or at all&mdash;related to these changes.  Adding a `nondeterministic` static field does not violate function purity, nor would its use by `FunctionGraph` features and optimizations.  If that were the case, then all features and optimizations that rely on `Op` type/class information and other non-`Apply` node inputs are in gross violation of this principle and are somehow "non-deterministic"&mdash;and that would quite a few of them.  

Furthermore, it's trivial to cast these ideas into a functional-like setting; simply introduce a state at the `Op`/type level.  If you want, consider it a state monad that's analogous to RNG state, except that it's at the graph level and only guarantees the uniqueness of nondeterministic `Op` outputs.  A concrete state monad probably isn't necessary, but it's perfectly do-able if need be.

Regardless, [as I mentioned before](https://github.com/Theano/Theano/pull/6691#issuecomment-469110172), discussions about this PR keep veering into vaguely related areas of concern.  Nothing about this PR is changing&mdash;or implying changes to&mdash;the functional determinism of Theano `Op`s.  It simply allows graph-processing functions to more easily treat "theoretically" distinct nodes as distinct by allowing one to designate an `Op` as _theoretically_ non-deterministic.  If functions use that information to produce different graphs, then that still doesn't affect the Theano purity contract: an `Op` still produces the same output given the same input.

Otherwise, If the concern is for _graph-level_ function purity, then I'm even more confused, because the "non-determinism" we're discussing isn't at the graph computation level.  A side-effect free function taking graphs with non-deterministic `Op`s as input will still produce the same outputs given equivalent inputs.

Reply:
> If Theano is designed to operate on that assumption, then it's setting an unnecessary and error-prone requirement/dependency between the graph and compute-levels—or its graphs are providing an extremely misleading representation of the relevant operations.

Yes, the computation graph relies on the assumption that the value in the inputs of on Op will not be changed when that Op gets evaluated (except if it can be guaranteed that no operation will ever read that value again, when inplace optimizations are introduced).
This implies that the value in a random state *must not change* during the evaluation of a function it is an input of. The way it is implemented when the underlying RNG implementation operates inplace is by copying the random state.

I understand it is not a natural way of seeing random operations, but I would not call the graph representation misleading: the graph representation is analogous to SSA, where values do not change once assign, and the un-ambiguous way of referring to "new" values of a state is to have the underlying operation ("sampling") explicitly return a new variable. The operations are explicitly deterministic given a state, and output a new state.

Or maybe "sampling" is the confusing term? I should clarify that in that context, each of the "sampling" functions in raw_random perform a deterministic transformation from a PRNG state (and other parameters) to a "sample" and an updated PRNG state. If you have a better term for this procedure, maybe we should use it for this discussion.

> Nothing about this PR is changing—or implying changes to—the functional determinism/purity of Theano Ops.

Your change seem to imply there would be non-deterministic operations in graph in the future, which violate the assumption above, and would have consequences reaching well beyond the "merge" optimization.

> Otherwise, If the concern is for _graph-level_ function purity, then I'm even more confused, because the "non-determinism" we're discussing isn't at the graph computation level.

Each Op in a graph should be pure, yes. The only computational side-effect happen at the boundaries of the graph, with the "updates" mechanism. This also ensures that different execution strategies of the same graph give the same input (modulo precision errors) regardless of execution order.